experimental/pattern-matching.md

```diff
@@ -0,0 +1,119 @@
+# [Pattern Matching](proposal-pattern-matching)
+
+See also [AST examples](ast-examples) for this proposal.
+
+## MatchExpression
+
+```js
+interface MatchExpression <: Expression {
+    type: "MatchExpression";
+    discriminant: Expression;
+    id: Pattern | null;
+    clause: [ MatchClause ]
+}
+```
+
+## WhenClause
+```js
+interface MatchClause <: Node {
+    type: "MatchClause";
+    test: MatchPattern | null;
+    guard: Expression | null;
+    id: Pattern | null;
+    consequent: BlockStatement
+}
+```
+
+If `test` is `null`, `id` must be `null`.
+
+## MatchPattern
```

Comment:
Could we add a mention here as to why we don’t just use Pattern? (You mentioned in the PR itself but seems important to include here as well.)

Reply:
Sure, addressed in d67b89a.